### DISCUSSION TITLE
Cyclone 5 Fpga data read write from HPS using Uboot c application

### DISCUSSION THREAD

**Discussion by Ravin-git** [community/community#145775] [Programming Help]
### Select Topic Area

Question

### Body

Need to read and write data from hps to fpga using uboot c application by using lwaxi bridge cyclone 5 socfpga

**Comment by mecodeatlas** [comment]
Thanks for posting in the GitHub Community, @Ravin-git!

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Discussions category is solely related to conversations around the [GitHub product Discussions](https://docs.github.com/en/discussions/collaborating-with-your-community-using-discussions/about-discussions). This question should be in the Programming Help category. I've gone ahead and moved it for you. Good luck!

**Comment by jgschmitz** [comment]
@Ravin-git  Ensure the device tree includes the LWAXI bridge node for the Cyclone V SoC:

```
lw_axi_bridge: bridge@0xff200000 {
    compatible = "altr,lwhps2fpga-bridge";
    reg = <0xff200000 0x10000>; /* Base address and size */
    resets = <&rst LW_AXI_BRIDGE_RESET>;
    clocks = <&clk LW_AXI_BRIDGE_CLK>;
};
```
Rebuild the U-Boot if necessary to include the updated device tree.

The LWAXI bridge is typically mapped at 0xFF200000 in HPS memory space. Confirm this in your FPGA design.

Create a standalone C application that runs in U-Boot to interact with the LWAXI bridge.

Example: Read and Write to LWAXI

```
#include <common.h>
#include <asm/io.h>

#define LWAXI_BASE_ADDR 0xFF200000  // Base address of the LWAXI bridge
#define OFFSET 0x00000010          // Example offset within the bridge

void lwaxi_test(void) {
    uint32_t write_data = 0xA5A5A5A5;  // Data to write
    uint32_t read_data;

    // Write to LWAXI
    writel(write_data, LWAXI_BASE_ADDR + OFFSET);
    printf("Wrote data: 0x%08X to address: 0x%08X\n", write_data, LWAXI_BASE_ADDR + OFFSET);

    // Read from LWAXI
    read_data = readl(LWAXI_BASE_ADDR + OFFSET);
    printf("Read data: 0x%08X from address: 0x%08X\n", read_data, LWAXI_BASE_ADDR + OFFSET);

    // Verify
    if (read_data == write_data) {
        printf("Read/Write operation successful!\n");
    } else {
        printf("Read/Write operation failed!\n");
    }
}
```

Cross-compile the C code using the ARM cross-compiler (arm-linux-gnueabihf-gcc):

```
arm-linux-gnueabihf-gcc -o lwaxi_test lwaxi_test.c
```
Transfer the compiled binary to the Cyclone V SoC, e.g., using TFTP or an SD card.

Run in U-Boot:
Boot the Cyclone V SoC into U-Boot.
Execute the application using the go command, providing the memory address of the binary

happy hacking - -
